lib/ProductOpener/Data.pm

```diff
@@ -60,45 +60,58 @@ sub execute_query {
 	my ($sub) = @_;
 
 	return Action::Retry->new(
-          attempt_code => sub { $action->run($sub) },
-          on_failure_code => sub { my ($error, $h) = @_; die $error; }, # by default Action::Retry would return undef
-		  # If we didn't get results from MongoDB, the server is probably overloaded
-		  # Do not retry the query, as it will make things worse
-		  strategy => { Fibonacci => { max_retries_number => 0, } },
-      )->run();
+		attempt_code => sub { $action->run($sub) },
+		on_failure_code => sub { my ($error, $h) = @_; die $error; }, # by default Action::Retry would return undef
+		# If we didn't get results from MongoDB, the server is probably overloaded
+		# Do not retry the query, as it will make things worse
+		strategy => { Fibonacci => { max_retries_number => 0, } },
+	)->run();
 }
 
 sub get_products_collection {
-	return get_collection($mongodb, 'products');
+	my ($timeout) = @_;
+	return get_collection($mongodb, 'products', $timeout);
 }
 
 sub get_products_tags_collection {
-	return get_collection($mongodb, 'products_tags');
+	my ($timeout) = @_;
+	return get_collection($mongodb, 'products_tags', $timeout);
 }
 
 sub get_emb_codes_collection {
-	return get_collection($mongodb, 'emb_codes');
+	my ($timeout) = @_;
+	return get_collection($mongodb, 'emb_codes', $timeout);
 }
 
 sub get_recent_changes_collection {
-	return get_collection($mongodb, 'recent_changes');
+	my ($timeout) = @_;
+	return get_collection($mongodb, 'recent_changes', $timeout);
 }
 
 sub get_collection {
-	my ($database, $collection) = @_;
-	return get_mongodb_client()->get_database($database)->get_collection($collection);
+	my ($database, $collection, $timeout) = @_;
+	return get_mongodb_client($timeout)->get_database($database)->get_collection($collection);
 }
 
 sub get_database {
-	return get_mongodb_client()->get_database($mongodb);
+	my ($database) = @_ // $mongodb;
+	return get_mongodb_client()->get_database($database);
 }
 
 sub get_mongodb_client() {
+	my ($timeout) = @_;
+
+	my %client_options = (
+		host => $mongodb_host,
+
+		# https://metacpan.org/pod/MongoDB::MongoClient#socket_timeout_ms
+		# default is 30000 ms
+		socket_timeout_ms => $timeout // 30000,
+	);
+
 	if (!defined($client)) {
-		$log->info("Creating new DB connection");
-			$client = MongoDB::MongoClient->new(
-				host => $mongodb_host
-			);
+		$log->info("Creating new DB connection, socket_timeout_ms ${client_options{socket_timeout_ms}}");
```

Comment:
You could use [structured logging](https://metacpan.org/pod/Log::Any#Logging) for the parameter. Log4net will format it into the same line when debugging, but if we ever move to an ELK stack or any similar system that supports structured logging, the value would be logged as a property.
```suggestion
		$log->info("Creating new DB connection", { socket_timeout_ms => $client_options{socket_timeout_ms} });
```

Reply:
interesting, thanks.